param/__init__.py

```diff
@@ -1838,6 +1842,9 @@ def _validate(self, val):
 
         self._checkBounds(val)
 
+    @classmethod
+    def deserialize(cls, value):
+        return dt.datetime.fromisoformat(value)
```

Comment:
Confused by this, either there should be a symmetric serialize method that outputs the iso formatted string or this should happen in the JSON decoder, which seems like the better solution.

Reply:
Serialize and deserialize should always support a round-trip:

```python
p = parameterized.param.value
parameterized.value = p.deserialize(p.serialize(parameterized.value))
```

Reply:
Round tripping will be tested but unfortunately Python's JSON `Encoder` and `Decoder` are *not* symmetric if you check the docs (see the example for encoding/decoding a complex number).

What I can do is move the logic out the encoder though which will make it symmetric for datetimes at least.

Reply:
Seems like we should attempt to support round tripping as best we can, to preserve our sanity and to support lots of potential workflows that depend on it.  So the fact that the classes aren't symmetric is annoying, but our syntax should be symmetric if at all possible, at least when using a schema during decoding. (Without a schema there's no hope, as e.g. we don't know whether "0.9" is meant to be a Magnitude or a Number...)

Reply:
Just to be very clear here, my requirement here is that I expect to use serialize/deserialize for other purposes, e.g. for URL parameter encoding and decoding. URL parameters are going to be much more limited than JSON so it will only work for a subset of types but I want to ensure that I can round-trip the few types that will be supported in a straightforward way and my hope that way is simply:

- serialize: `html.escape(Parameter.serialize(value))`
- deserialize `Parameter.deserialize(html.unescape(value))`

Reply:
I think with the IR approach, the `serialize` and `deserialize` methods should always be symmetric as you say. Only when we need a custom `JSONEncoder` does it get tricky (and so far it has been possible to avoid).